### PR TITLE
GH-37229: [MATLAB] Add `arrow.type.Date32Type` class and `arrow.date32` construction function

### LINKED ISSUE #37229: [MATLAB] Add `arrow.type.Date32Type` class and `arrow.date32` construction function
### Describe the enhancement requested

To add support for `arrow.array.Date32Array`, we will need to add support for a `arrow.type.Date32Type` class and associated `arrow.date32` construction function to the MATLAB interface.

### Component(s)

MATLAB

### PR DESCRIPTION
**Pull request by kevingurney**
### Rationale for this change

In support of adding `arrow.array.Date32Array`, this pull request adds a new `arrow.type.Date32Type` class and associated `arrow.date32` construction function to the MATLAB interface.

### What changes are included in this PR?

1. New `arrow.type.Date32Type` class.
2. New `arrow.date32` construction function that returns an `arrow.type.Date32Type` instance.
3. New `arrow.type.ID.Date32` type enumeration value.
4. Added an abstract `arrow.type.DateType` class which `arrow.type.Date32Type` inherits from. `arrow.type.Date64Type` will also inherit from this class to share the `DateUnit` property. This mirrors the implementation of the `Time32Type` and `Time64Type` classes inheriting from an abstract `arrow.type.TimeType` class to share the `TimeUnit` property.

**Example**

```matlab
>> type = arrow.date32()

type = 

  Date32Type with properties:

          ID: Date32
    DateUnit: Day
```

### Are these changes tested?

Yes.

1. Added a new `tDate32Type` test class.
2. Updated the `tID` test class to include `arrow.type.ID.Date32`. 

### Are there any user-facing changes?

Yes.

1. There is a new public `arrow.type.Date32Type` class.
2. There is a new public `arrow.date32` construction function.
3. There is a new `arrow.type.ID.Date32` type enumeration value.

### Future Directions

1. #37230
2. Add `arrow.array.Date32Array` class.
3. Add `arrow.array.Date64Array` class.
* Closes: #37229

### CODE REVIEW

File: matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/matlab/type/proxy/date32_type.h"
+#include "arrow/matlab/error/error.h"
+#include "arrow/util/utf8.h"
```
Comment (sgilmore10) [inline matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc:20]: You can delete `#include arrow/util/utf8.h`. I don't think we're using any utilities included in that header here.

File: matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/matlab/type/proxy/date32_type.h"
+#include "arrow/matlab/error/error.h"
```
Comment (sgilmore10) [inline matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc:19]: I think you can delete `#include "arrow/matlab/error/error.h"`

File: matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+#include "arrow/util/utf8.h"
+
+namespace arrow::matlab::type::proxy {
+
+    Date32Type::Date32Type(std::shared_ptr<arrow::Date32Type> date32_type) : DateType(std::move(date32_type)) {}
+
+    libmexclass::proxy::MakeResult Date32Type::make(const libmexclass::proxy::FunctionArguments& constructor_arguments) {
+        namespace mda = ::matlab::data;
```
Comment (sgilmore10) [inline matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc:27]: You can delete `namespace mda = ::matlab::data`. I don't think you use it below.

File: matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc (hunk 1 above)
Comment (kevingurney) [reply matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc:20]: Good catch. Sorry, that was left over from following the `time32_type.cc` code as a template, where we were converting strings.

File: matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc (hunk 3 above)
Comment (kevingurney) [reply matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc:27]: Good catch!

**Review by sgilmore10 — CHANGES_REQUESTED** [review]
Looks really good! Just found a few typos.

File: matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc (hunk 1 above)
Comment (kevingurney) [reply matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc]: Fixed.

File: matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc (hunk 3 above)
Comment (kevingurney) [reply matlab/src/cpp/arrow/matlab/type/proxy/date32_type.cc]: Fixed.

**Review by sgilmore10 — APPROVED** [review]

**Comment by kevingurney** [comment]
+1

**Pull request merged**